File: src/ru/hudspeed/pro/RadarBaseUpdater.java
package ru.gpsantiradar.app;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.GZIPInputStream;

public final class RadarBaseUpdater {
    public static final String ACTION_DATABASE_UPDATED =
            "ru.gpsantiradar.app.RADARBASE_UPDATED";

    private static final String RADARBASE_URL =
            "https://radarbase.info/export/cache/RU/main_extended.json";
    private static final String RADARBASE_ETAG = "radarbase_etag";
    private static final String RADARBASE_MODIFIED = "radarbase_modified";
    private static final String RADARBASE_COORDINATE_FIX = "radarbase_coordinate_fix_v1";
    private static final String RADARBASE_IMPORT_FORMAT = "radarbase_import_format";
    private static final int CURRENT_RADARBASE_IMPORT_FORMAT = 2;

    public interface Listener {
        void onRadarBaseUpdate(RadarBaseUpdateState state);
    }

    private final Context context;
    private final RadarBaseUpdateSingleFlight gate;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final List<Listener> listeners = new ArrayList<>();
    private volatile RadarBaseUpdateState latestState = new RadarBaseUpdateState(
            0, RadarBaseUpdateState.Status.IDLE, 0, false, "");
    private long sequence;

    public RadarBaseUpdater(Context context, RadarBaseUpdateSingleFlight gate) {
        this.context = context.getApplicationContext();
        this.gate = gate;
    }

    public void requestUpdate() {
        if (!gate.tryStart()) {
            publish(RadarBaseUpdateState.Status.ALREADY_RUNNING, 0, false,
                    "Обновление базы RadarBase уже выполняется");
            return;
        }
        publish(RadarBaseUpdateState.Status.STARTED, 0, false,
                "Обновление базы RadarBase начато");
        new Thread(new Runnable() {
            @Override public void run() {
                updateInBackground();
            }
        }, "radarbase-download").start();
    }

    public RadarBaseUpdateState latestState() {
        return latestState;
    }

    public synchronized void addListener(Listener listener, boolean replayLatest) {
        if (listener == null) return;
        if (!listeners.contains(listener)) listeners.add(listener);
        if (replayLatest) {
            List<Listener> replay = new ArrayList<>();
            replay.add(listener);
            dispatch(replay, latestState);
        }
    }

    public synchronized void removeListener(Listener listener) {
        listeners.remove(listener);
    }

    private void updateInBackground() {
        HttpURLConnection connection = null;
        try {
            SharedPreferences preferences = context.getSharedPreferences(
                    AppSettings.PREFERENCES, Context.MODE_PRIVATE);
            prepareMigration(preferences);
            connection = (HttpURLConnection) new URL(RADARBASE_URL).openConnection();
            connection.setConnectTimeout(15000);
            connection.setReadTimeout(120000);
            connection.setRequestProperty("Accept", "application/json");
            connection.setRequestProperty("Accept-Encoding", "gzip");
            String etag = preferences.getString(RADARBASE_ETAG, "");
            String modified = preferences.getString(RADARBASE_MODIFIED, "");
            if (etag != null && !etag.isEmpty()) {
                connection.setRequestProperty("If-None-Match", etag);
            }
            if (modified != null && !modified.isEmpty()) {
                connection.setRequestProperty("If-Modified-Since", modified);
            }

            int status = connection.getResponseCode();
            if (status == HttpURLConnection.HTTP_NOT_MODIFIED) {
                publish(RadarBaseUpdateState.Status.UNCHANGED, 0, false,
                        "База RadarBase уже актуальна");
                return;
            }
            if (status != HttpURLConnection.HTTP_OK) {
                throw new IOException("Сервер RadarBase: HTTP " + status);
            }

            InputStream raw = connection.getInputStream();
            InputStream input = "gzip".equalsIgnoreCase(connection.getContentEncoding())
                    ? new GZIPInputStream(raw) : raw;
            RadarBaseParser.Result result;
            try (InputStream source = input;
                 CameraDatabase db = new CameraDatabase(context)) {
                result = db.importRadarBase(source);
            }

            SharedPreferences.Editor editor = preferences.edit();
            String newEtag = connection.getHeaderField("ETag");
            String newModified = connection.getHeaderField("Last-Modified");
            if (newEtag != null) editor.putString(RADARBASE_ETAG, newEtag);
            if (newModified != null) editor.putString(RADARBASE_MODIFIED, newModified);
            editor.putBoolean(RADARBASE_COORDINATE_FIX, true);
            editor.putInt(RADARBASE_IMPORT_FORMAT, CURRENT_RADARBASE_IMPORT_FORMAT);
            editor.putLong(AppSettings.RADARBASE_LAST_SUCCESSFUL_DOWNLOAD,
                    System.currentTimeMillis());
            editor.apply();

            context.sendBroadcast(
                    new Intent(ACTION_DATABASE_UPDATED).setPackage(context.getPackageName()));
            String suffix = result.coordinatesCorrected ? " · координаты восстановлены"
                    : " · без поправки координат";
            publish(RadarBaseUpdateState.Status.SUCCESS, result.count,
                    result.coordinatesCorrected, "Импортировано: " + result.count + suffix);
        } catch (Exception error) {
            String detail = error.getMessage();
            publish(RadarBaseUpdateState.Status.ERROR, 0, false,
                    detail == null || detail.trim().isEmpty()
                            ? "Не удалось загрузить JSON RadarBase" : detail);
        } finally {
            if (connection != null) connection.disconnect();
            gate.finish();
        }
    }

    private static void prepareMigration(SharedPreferences preferences) {
        boolean coordinatesReady = preferences.getBoolean(RADARBASE_COORDINATE_FIX, false);
        boolean importReady = preferences.getInt(RADARBASE_IMPORT_FORMAT, 0)
                >= CURRENT_RADARBASE_IMPORT_FORMAT;
        if (!coordinatesReady || !importReady) {
            preferences.edit().remove(RADARBASE_ETAG).remove(RADARBASE_MODIFIED).commit();
        }
    }

    private void publish(RadarBaseUpdateState.Status status, int importedCount,
                         boolean coordinatesCorrected, String message) {
        RadarBaseUpdateState state;
        List<Listener> snapshot;
        synchronized (this) {
            state = new RadarBaseUpdateState(++sequence, status, importedCount,
                    coordinatesCorrected, message);
            latestState = state;
            snapshot = new ArrayList<Listener>(listeners);
        }
        dispatch(snapshot, state);
    }

    private void dispatch(final List<Listener> targets, final RadarBaseUpdateState state) {
        mainHandler.post(new Runnable() {
            @Override public void run() {
                for (Listener listener : targets) {
                    listener.onRadarBaseUpdate(state);
                }
            }
        });
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/RadarBaseUpdater.kt
package ru.gpsantiradar.app

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.zip.GZIPInputStream

class RadarBaseUpdater(context: Context, private val gate: RadarBaseUpdateSingleFlight) {
    interface Listener : RadarBaseUpdateListenerRegistry.Listener
    private val context = context.applicationContext
    private val listenerRegistry: RadarBaseUpdateListenerRegistry

    init {
        val mainHandler = Handler(Looper.getMainLooper())
        listenerRegistry = RadarBaseUpdateListenerRegistry { callback -> mainHandler.post(callback) }
    }

    fun requestUpdate() {
        if (!gate.tryStart()) {
            publish(RadarBaseUpdateState.Status.ALREADY_RUNNING, 0, false, "Обновление базы RadarBase уже выполняется")
            return
        }
        publish(RadarBaseUpdateState.Status.STARTED, 0, false, "Обновление базы RadarBase начато")
        Thread(::updateInBackground, "radarbase-download").start()
    }
    fun latestState() = listenerRegistry.latestState()
    fun addListener(listener: Listener?, replayLatest: Boolean) = listenerRegistry.addListener(listener, replayLatest)
    fun removeListener(listener: Listener?) = listenerRegistry.removeListener(listener)

    private fun updateInBackground() {
        var connection: HttpURLConnection? = null
        try {
            val preferences = context.getSharedPreferences(AppSettings.PREFERENCES, Context.MODE_PRIVATE)
            prepareMigration(preferences)
            connection = URL(RADARBASE_URL).openConnection() as HttpURLConnection
            connection.connectTimeout = 15000
            connection.readTimeout = 120000
            connection.setRequestProperty("Accept", "application/json")
            connection.setRequestProperty("Accept-Encoding", "gzip")
            preferences.getString(RADARBASE_ETAG, "")?.takeIf { it.isNotEmpty() }?.let { connection.setRequestProperty("If-None-Match", it) }
            preferences.getString(RADARBASE_MODIFIED, "")?.takeIf { it.isNotEmpty() }?.let { connection.setRequestProperty("If-Modified-Since", it) }
            val status = connection.responseCode
            if (status == HttpURLConnection.HTTP_NOT_MODIFIED) {
                publish(RadarBaseUpdateState.Status.UNCHANGED, 0, false, "База RadarBase уже актуальна")
                return
            }
            if (status != HttpURLConnection.HTTP_OK) throw IOException("Сервер RadarBase: HTTP $status")
            val raw = connection.inputStream
            val input = if (connection.contentEncoding.equals("gzip", true)) GZIPInputStream(raw) else raw
            val result = input.use { source -> CameraDatabase(context).use { it.importRadarBase(source) } }
            preferences.edit().apply {
                connection.getHeaderField("ETag")?.let { putString(RADARBASE_ETAG, it) }
                connection.getHeaderField("Last-Modified")?.let { putString(RADARBASE_MODIFIED, it) }
                putBoolean(RADARBASE_COORDINATE_FIX, true)
                putInt(RADARBASE_IMPORT_FORMAT, CURRENT_RADARBASE_IMPORT_FORMAT)
                putLong(AppSettings.RADARBASE_LAST_SUCCESSFUL_DOWNLOAD, System.currentTimeMillis())
            }.apply()
            context.sendBroadcast(Intent(ACTION_DATABASE_UPDATED).setPackage(context.packageName))
            val suffix = if (result.coordinatesCorrected) " · координаты восстановлены" else " · без поправки координат"
            publish(RadarBaseUpdateState.Status.SUCCESS, result.count, result.coordinatesCorrected, "Импортировано: ${result.count}$suffix")
        } catch (error: Exception) {
            publish(RadarBaseUpdateState.Status.ERROR, 0, false,
                error.message?.takeIf { it.isNotBlank() } ?: "Не удалось загрузить JSON RadarBase")
        } finally {
            connection?.disconnect()
            gate.finish()
        }
    }

    private fun publish(status: RadarBaseUpdateState.Status, importedCount: Int, coordinatesCorrected: Boolean, message: String) {
        listenerRegistry.publish(status, importedCount, coordinatesCorrected, message)
    }

    companion object {
        const val ACTION_DATABASE_UPDATED = "ru.gpsantiradar.app.RADARBASE_UPDATED"
        private const val RADARBASE_URL = "https://radarbase.info/export/cache/RU/main_extended.json"
        private const val RADARBASE_ETAG = "radarbase_etag"
        private const val RADARBASE_MODIFIED = "radarbase_modified"
        private const val RADARBASE_COORDINATE_FIX = "radarbase_coordinate_fix_v1"
        private const val RADARBASE_IMPORT_FORMAT = "radarbase_import_format"
        private const val CURRENT_RADARBASE_IMPORT_FORMAT = 2
        private fun prepareMigration(preferences: SharedPreferences) {
            val coordinatesReady = preferences.getBoolean(RADARBASE_COORDINATE_FIX, false)
            val importReady = preferences.getInt(RADARBASE_IMPORT_FORMAT, 0) >= CURRENT_RADARBASE_IMPORT_FORMAT
            if (!coordinatesReady || !importReady) preferences.edit().remove(RADARBASE_ETAG).remove(RADARBASE_MODIFIED).commit()
        }
    }
}
